storage/src/journal/fixed.rs

```diff
@@ -153,32 +153,62 @@ impl<E: Storage + Metrics, A: Codec<Cfg = ()> + FixedSize> Journal<E, A> {
         context.register("pruned", "Number of blobs pruned", pruned.clone());
         tracked.set(blobs.len() as i64);
 
-        // truncate the last blob if it's not the expected length, which might happen from unclean
+        // Truncate the last blob if it's not the expected length, which might happen from unclean
         // shutdown.
         let newest_blob_index = *blobs.keys().last().unwrap();
-        let (newest_blob, len) = blobs.get_mut(&newest_blob_index).unwrap();
-        if *len % Self::CHUNK_SIZE_U64 != 0 {
+        let newest_blob = blobs.get_mut(&newest_blob_index).unwrap();
+        let mut size = newest_blob.size().await;
+        if size % Self::CHUNK_SIZE_U64 != 0 {
             warn!(
                 blob = newest_blob_index,
-                invalid_len = *len,
-                "last blob len is not a multiple of item size, truncating"
+                invalid_size = size,
+                "last blob size is not a multiple of item size, truncating"
             );
-            *len -= *len % Self::CHUNK_SIZE_U64;
-            newest_blob.truncate(*len).await?;
+            size -= size % Self::CHUNK_SIZE_U64;
+            newest_blob.truncate(size).await?;
             newest_blob.sync().await?;
         }
 
-        if *len == cfg.items_per_blob * Self::CHUNK_SIZE_U64 {
+        // Truncate any records with failing checksums. This can happen if the file system allocated
+        // extra space for a blob but there was a crash before any data was written to that space.
+        let mut truncated = false;
+        while size > 0 {
```

Comment:
We should never need to backtrack across multiple journals because we will never create a new one without first syncing the previous.